core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java

```diff
@@ -49,12 +50,12 @@
   private long killTime = 1000 * 3;
 
   private static class CachedConnections {
-    LinkedList<CachedConnection> unreserved = new LinkedList<>();
-    Map<CachedTTransport,CachedConnection> reserved = new HashMap<>();
+    Deque<CachedConnection> unreserved = new LinkedBlockingDeque<>(); // stack - LIFO
+    Map<CachedTTransport,CachedConnection> reserved = new ConcurrentHashMap<>();
 
     public CachedConnection reserveAny() {
       if (unreserved.size() > 0) {
-        CachedConnection cachedConnection = unreserved.removeFirst();
+        CachedConnection cachedConnection = unreserved.pop();
```

Comment:
This code now has a race condition.  After checking if size > 0, Dequeue could become empty before calling pop.  There is a method called pollFirst that could be used to do all of this in a single operation.  However I am not 100% sure that pop and pollFirst take from the same end.

```java
      CachedConnection cachedConnection = unreserved.pollFirst();
      if (cachedConnection != null) {
        cachedConnection.reserve();
        reserved.put(cachedConnection.transport, cachedConnection);
        if (log.isTraceEnabled()) {
          log.trace("Using existing connection to {}", cachedConnection.transport.cacheKey);
        }
        return cachedConnection;
      }
```